modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/PageMemoryMvPartitionStorage.java

```diff
@@ -0,0 +1,515 @@
+        VersionChain currentChain = findVersionChainForModification(rowId);
+
+        throwIfChainBelongsToAnotherTx(currentChain, txId);
+
+        RowVersion currentVersion = findLatestRowVersion(currentChain);
+        RowVersion newVersion = insertRowVersion(row, currentVersion.isUncommitted() ? currentVersion.nextLink() : currentChain.headLink());
+
+        if (currentVersion.isUncommitted()) {
+            // as we replace an uncommitted version with new one, we need to remove old uncommitted version
+            removeRowVersion(currentVersion);
+        }
+
+        VersionChain chainReplacement = new VersionChain(
+                partitionId,
+                txId,
+                PartitionlessLinks.removePartitionIdFromLink(newVersion.link())
+        );
+
+        updateVersionChain(currentChain, chainReplacement);
+
+        if (currentVersion.isUncommitted()) {
+            return rowVersionToBinaryRow(currentVersion);
+        } else {
+            return null;
+        }
+    }
+
+    @NotNull
+    private VersionChain findVersionChainForModification(RowId rowId) {
+        VersionChain currentChain = findVersionChain(rowId);
+        if (currentChain == null) {
+            throw new RowIdIsInvalidForModificationsException();
+        }
+        return currentChain;
+    }
+
+    @Override
+    public @Nullable BinaryRow abortWrite(RowId rowId) throws StorageException {
+        VersionChain currentVersionChain = findVersionChainForModification(rowId);
+
+        if (currentVersionChain.transactionId() == null) {
+            throw new NoUncommittedVersionException();
+        }
+
+        RowVersion currentVersion = findLatestRowVersion(currentVersionChain);
+        assert currentVersion.isUncommitted();
+
+        removeRowVersion(currentVersion);
+
+        if (currentVersion.hasNextLink()) {
+            VersionChain versionChainReplacement = VersionChain.withoutTxId(
+                    partitionId,
+                    currentVersionChain.link(),
+                    currentVersion.nextLink()
+            );
+            updateVersionChain(currentVersionChain, versionChainReplacement);
+        } else {
+            // it was the only version, let's remove the chain as well
+            removeVersionChain(currentVersionChain);
+        }
+
+        return rowVersionToBinaryRow(currentVersion);
+    }
+
+    private void removeVersionChain(VersionChain currentVersionChain) {
+        try {
+            versionChainFreeList.removeDataRowByLink(currentVersionChain.link());
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Cannot remove chain version", e);
+        }
+    }
+
+    @Override
+    public void commitWrite(RowId rowId, Timestamp timestamp) throws StorageException {
+        VersionChain currentVersionChain = findVersionChainForModification(rowId);
```

Comment:
I don't think that "VersionChain" is a good name for what we have. It's more like a descriptor, a head. Chain is what it refers to. Also, it's easy to confuse with "Versioned Chain", the name that's used in many sources about MV storages.

Reply:
Well, it can be regarded as a lazy chain, or like a reference to a chain (and we usually do not distinguish references to objects and the objects themselves). When in Java you have a `LinkedList` object, its fields actually only contain reference to next items, not the whole list content, as `ArrayList` does. With `VersionChain`, the only difference is that it does not 'lazy-load' itself, but the 'dereference' is done using an external (to `VersionChain`) code. So it seems acceptable.

As 'Versioned Chain', I did not find such a term, could you please point me to the corresponding documents? But, googling by 'Version Chain MVCC', I found pages talking about Version Chains, which seem to be the same thing that is denoted by `VersionChain` in this PR. So it does not seem to be too confusing...

Anyway, if you still don't like it, what alternative name would you propose?

Reply:
We had a conversation about this. There are pros and cons in each of the proposed naming schemes (VersionChain/RowVersion, VersionHistory/VersionChain, etc), so for now I leave it as is, we'll be able to revisit the naming later. I added a short clarification to the class javadoc.